### PR TITLE
Initial implementation of IpNetwork/IpInterface for hack-session

### PR DESCRIPTION
**Pull request by jparyani**

### CODE REVIEW

File: shell/server/mail.js [hunk 1]
```diff
@@ -535,6 +538,256 @@ HackSessionContextImpl.prototype.revokeApiToken = function (tokenId) {
   }).bind(this));
 }
 
+function IpInterfaceImpl () { }
+
+IpInterfaceImpl.prototype.listenTcp = function (portNum, port) {
+  return new Promise(function (resolve, reject) {
+    var resolved = false;
+    var server = Net.createServer(function (connection) {
+      connection.done = connection.end;
```
Comment (kentonv) [inline shell/server/mail.js]: This is a little too clever for me. :)

Can we explicitly define an object here meant to implement the Cap'n Proto interface?

I worry about things like we add a new method to ByteStream which happens to conflict with a method in node's connection type, and then suddenly things inexplicably don't work.

File: shell/server/mail.js [hunk 2]
```diff
@@ -535,6 +538,256 @@ HackSessionContextImpl.prototype.revokeApiToken = function (tokenId) {
...
+
+};
+
+function BoundUdpPortImpl(server, address, port) {
+  this.server = server;
+  this.address = address;
+  this.port = port;
+}
+
+BoundUdpPortImpl.prototype.send = function(message, returnPort) {
```
Comment (kentonv) [inline shell/server/mail.js]: Comment that we're assuming that `returnPort` is always the one corresponding to `this.server`, but that we should really be maintaining a map, but that that requires capability equality which we don't have yet.

File: shell/server/mail.js [hunk 3]
```diff
@@ -535,6 +538,256 @@ HackSessionContextImpl.prototype.revokeApiToken = function (tokenId) {
...
+    // start with lower bits of address and build the output in reverse
+    // this ensures that we can place a colon every 4 characters
+    out += hex[hex.length - 1 - i];
+    if ((i + 1) % 4 === 0) {
+      out += ':';
+      ++numColons;
+    }
+  }
+
+  // need to fill out colons on the upper end of the address til we have 7
```
Comment (kentonv) [inline shell/server/mail.js]: I think a double-colon can be used to mean "fill with as many zeroes as necessary" in an ipv6 address... though I might be making that up.

**Comment by kentonv** [comment]
Can we put all this IP code in a separate file? `mail.js` is getting a little overloaded. :) We should also refactor the non-mail stuff out of mail.js, in a separate change...

**Comment by jparyani** [comment]
All your comments are now fixed.

**Pull request merged**